### PR TITLE
Tekufots and prayers calculation

### PR DESCRIPTION
**Pull request by DanBendavid**
new file:   hdate/tekufot.py
	new file:   tests/test_tekufot.py

# Description
The class attempts to compute:
    - The tekufot (seasons) in the Hebrew calendar: Nissan, Tammuz, Tishrei, and Tevet.
    - Cheilat Geshamim start date, which differs between the diaspora and Israel.
    - Halachic prayer periods based on key Jewish holidays and seasonal changes.
    - Appropriate prayer phrases depending on the current date, tradition, and language.
"""
# Closes issue(s)

  Fixes: #<issue number>

# Changes included (select only one)

- [ ] Bugfix (non-breaking change that solves an issue)
- [X] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible)

# Checklist

- [ ] Code passes tox

### CODE REVIEW

File: hdate/tekufot.py [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+    ):
+        """
+        Calculates the start date for the prayers for rain (Cheilat Geshamim).
+        In the diaspora, it is 60 days (add 59 days) after Tekufat Tishrei.
+        In Israel, it is fixed at the 7th of Cheshvan.
+        """
+        from hdate import Zmanim
+
+        # Ensure we have Tekufa Tishrei
+        if not hasattr(self, "tekufa_tishrei") or self.tekufa_tishrei is None:
```
Comment (tsvi) [inline hdate/tekufot.py:208]: Could this be?

File: hdate/tekufot.py [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+            "tekufa_tevet": getattr(self, "tekufa_tevet", None),
+            "tekufa_nissan": self.tekufa_nissan,
+            "tekufa_tammuz": getattr(self, "tekufa_tammuz", None),
+            "cheilat_geshamim": getattr(self, "cheilat_geshamim", None),
+            "seventh_cheshvan": getattr(self, "seventh_cheshvan", None),
+            "prev_pesach": getattr(self, "prev_pesach", None),
+            "shemini_atzeret": getattr(self, "shemini_atzeret", None),
+            "pesach": getattr(self, "pesach", None),
+            "next_shemini_atzeret": getattr(self, "next_shemini_atzeret", None),
+        }
```
Comment (tsvi) [inline hdate/tekufot.py:346]: If all results are part of self, there's no need for this method.

File: tests/test_tekufot.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+            assert isinstance(period[0], str)
+            assert isinstance(period[1], datetime.date)
+            assert isinstance(period[2], datetime.date)
+
+    def test_is_leap(self):
+        """Test is_leap method with known leap and common years."""
+        assert Tekufot.is_leap(2020) is True
+        assert Tekufot.is_leap(2100) is False
+        assert Tekufot.is_leap(2000) is True
+        assert Tekufot.is_leap(2021) is False
```
Comment (tsvi) [inline tests/test_tekufot.py:122]: No need since isleap is builtin

**Review by tsvi — CHANGES_REQUESTED** [review]
Thank you very much.

These are a few comments.
Please address these as well as fix the errors, and merge issues. Then we can continue.

File: hdate/tekufot.py [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+        # Cheilat Geshamim calculation
+        self.get_cheilat_geshamim()
+
+        # Prayer periods
+        self.set_prayer_periods()
+
+    @staticmethod
+    def is_leap(year: int) -> bool:
+        """Determine if a given Gregorian year is a leap year."""
+        return year % 4 == 0 and (year % 100 != 0 or year % 400 == 0)
```
[the remark this answers is not in the document]
Comment (DanBendavid) [reply hdate/tekufot.py:155]: agree - i am not using it actually. so i removed this function.

File: hdate/tekufot.py (hunk 2 above)
Comment (DanBendavid) [reply hdate/tekufot.py:346]: i will remove it after . it is for debugging

File: hdate/tekufot.py (hunk 1 above)
Comment (DanBendavid) [reply hdate/tekufot.py]: ok - testing the Pylint

File: hdate/tekufot.py [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+                    cheilat_geshamim_dt.year,
+                    cheilat_geshamim_dt.month,
+                    cheilat_geshamim_dt.day,
+                )
+            else:
+                self.cheilat_geshamim = datetime.date(
+                    cheilat_geshamim_dt.year,
+                    cheilat_geshamim_dt.month,
+                    cheilat_geshamim_dt.day + 1,
+                )
```
[the remark this answers is not in the document]
Comment (DanBendavid) [reply hdate/tekufot.py]: i have added datetime

File: hdate/tekufot.py [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+                tzinfo=tz,
+            )
+
+            if cheilat_geshamim_dt < time_end_of_day:
+                # Normalize to date at midnight
+                self.cheilat_geshamim = datetime.date(
+                    cheilat_geshamim_dt.year,
+                    cheilat_geshamim_dt.month,
+                    cheilat_geshamim_dt.day,
+                )
```
[the remark this answers is not in the document]
Comment (DanBendavid) [reply hdate/tekufot.py]: ok

File: hdate/tekufot.py [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+from typing import Optional, Union
+
+from hdate import converters as conv
+from hdate.hebrew_date import HebrewDate
+from hdate.location import Location
+
+LANG = namedtuple("LANG", ["french", "english", "hebrew"])
+TRADITION = namedtuple(
+    "TRADITION", ["israel", "diaspora_ashkenazi", "diaspora_sephardi"]
+)
```
[the remark this answers is not in the document]
Comment (DanBendavid) [reply hdate/tekufot.py:21]: will work on this one

File: hdate/tekufot.py [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+
+    def get_cheilat_geshamim(
+        self,
+    ):
+        """
+        Calculates the start date for the prayers for rain (Cheilat Geshamim).
+        In the diaspora, it is 60 days (add 59 days) after Tekufat Tishrei.
+        In Israel, it is fixed at the 7th of Cheshvan.
+        """
+        from hdate import Zmanim
```
[the remark this answers is not in the document]
Comment (DanBendavid) [reply hdate/tekufot.py]: done

**Comment by DanBendavid** [comment]
I have many error from pylint. if you can guide me here.
Thanks 

hdate/tekufot.py:193: note: Use "-> None" if function does not return a value
hdate/tekufot.py:229: error: Incompatible types in assignment (expression has type "date", variable has type "datetime")  [assignment]
hdate/tekufot.py:231: error: Incompatible types in assignment (expression has type "date", variable has type "datetime")  [assignment]
hdate/tekufot.py:238: error: Incompatible types in assignment (expression has type "date", variable has type "datetime")  [assignment]
tests/test_tekufot.py:26: error: Function is missing a type annotation for one or more arguments  [no-untyped-def]
tests/test_tekufot.py:41: error: Function is missing a return type annotation  [no-untyped-def]
tests/test_tekufot.py:52: error: Function is missing a return type annotation  [no-untyped-def]
tests/test_tekufot.py:63: error: Function is missing a return type annotation  [no-untyped-def]
tests/test_tekufot.py:73: error: Function is missing a return type annotation  [no-untyped-def]
tests/test_tekufot.py:90: error: Function is missing a return type annotation  [no-untyped-def]
tests/test_tekufot.py:106: error: Function is missing a return type annotation  [no-untyped-def]

**Comment by DanBendavid** [comment]
i have added more test cases. Please check accuracy.

**Comment by tsvi** [comment]
> I have many error from pylint. if you can guide me here. Thanks
> 
> hdate/tekufot.py:193: note: Use "-> None" if function does not return a value hdate/tekufot.py:229: error: Incompatible types in assignment (expression has type "date", variable has type "datetime") [assignment] hdate/tekufot.py:231: error: Incompatible types in assignment (expression has type "date", variable has type "datetime") [assignment] hdate/tekufot.py:238: error: Incompatible types in assignment (expression has type "date", variable has type "datetime") [assignment] tests/test_tekufot.py:26: error: Function is missing a type annotation for one or more arguments 

**These should probably return None**
> [no-untyped-def] tests/test_tekufot.py:41: error: Function is missing a return type annotation [no-untyped-def] tests/test_tekufot.py:52: error: Function is missing a return type annotation [no-untyped-def] tests/test_tekufot.py:63: error: Function is missing a return type annotation [no-untyped-def] tests/test_tekufot.py:73: error: Function is missing a return type annotation [no-untyped-def] tests/test_tekufot.py:90: error: Function is missing a return type annotation [no-untyped-def] tests/test_tekufot.py:106: error: Function is missing a return type annotation [no-untyped-def]



> I have many error from pylint. if you can guide me here. Thanks
> 
> hdate/tekufot.py:193: note: Use "-> None" if function does not return a value hdate/tekufot.py:229: error: Incompatible types in assignment (expression has type "date", variable has type "datetime") [assignment] hdate/tekufot.py:231: error: Incompatible types in assignment (expression has type "date", variable has type "datetime") [assignment] hdate/tekufot.py:238: error: Incompatible types in assignment (expression has type "date", variable has type "datetime") [assignment] tests/test_tekufot.py:26: error: Function is missing a type annotation for one or more arguments [no-untyped-def] tests/test_tekufot.py:41: error: Function is missing a return type annotation [no-untyped-def] tests/test_tekufot.py:52: error: Function is missing a return type annotation [no-untyped-def] tests/test_tekufot.py:63: error: Function is missing a return type annotation [no-untyped-def] tests/test_tekufot.py:73: error: Function is missing a return type annotation [no-untyped-def] tests/test_tekufot.py:90: error: Function is missing a return type annotation [no-untyped-def] tests/test_tekufot.py:106: error: Function is missing a return type annotation [no-untyped-def]

You have in line 121 `self.cheilat_geshamim = tekufat_tishrei`, so the type is a datetime, but later on you assign to it a date. Technically it works, but it might hide a bug, that's why we don't like it.
I suppose you meant tekufat_tishrei.date(), or else you need to assigne date times further in the code.

**Review by tsvi — CHANGES_REQUESTED** [review]
Please request review only when everything is ready to be reviewed. I see your comments so you can always ask questions.
Ready for review means everything is fixed and should move on to be accepted.

**Comment by tsvi** [comment]
You don't have to close the PR. Just don't click ready for review, or request review.

**Comment by DanBendavid** [comment]
I ve replaced with a much simpler way see #PR161

**Pull request closed without merging**